### PR TITLE
Communication service

### PR DESCRIPTION
**Pull request by Sparsh-Kumar**
## Description
Added communication service to allow backend to send emails and sms.

## Database schema changes
```None```

### Automated test cases added
- should not send email, if recipient email or sender email is invalid or sender name is empty.
- should send email, if recipient email, sender email, sender name are valid.
- should not send sms, if recipient phone number is invalid or sms body is empty.
- should send sms, if recipient phone number is valid and sms body is not empty.

### Manual test cases run
```None```

### CODE REVIEW

File: src/modules/communication/internals/constant.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+export default emailRegex;
```
Comment (Sparsh-Kumar) [inline src/modules/communication/internals/constant.ts:2]: This cannot be achieved in 1 line like ```export default emailRegex = ...```

File: config/default.yml [hunk 2]
```diff
@@ -1,2 +1,4 @@
 jwt:
   expiresIn: "1y" # default token expire time 1 year. Examples - https://github.com/vercel/ms#examples
+
+webAppHost: "WEB_APP_HOST"
```
Comment (jjalan) [inline config/default.yml:4]: Please remove it from here as there is no default value for a web host. It varies per env.

File: config/development.yml [hunk 3]
```diff
@@ -0,0 +1 @@
+webAppHost: "WEB_APP_HOST"
```
Comment (jjalan) [inline config/development.yml:1]: Set it to realistic value - say `http://localhost:8000` (please confirm with @vindru  on which port does frontend would run)

File: src/modules/communication/internals/sendgrid-service.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      dynamicTemplateData: {
+        ...params.templateData,
+        domain: ConfigService.getStringValue('webAppHost'),
+      },
+    };
+
+    try {
+      await mail.send(msg);
+    } catch (e) {
+      throw new ThirdPartyServiceError('Email service unavailable.');
```
Comment (jjalan) [inline src/modules/communication/internals/sendgrid-service.ts:31]: Please log the third party error message so we can get notified and proactively investigate about it.

File: src/modules/communication/internals/twilio-params.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  ValidationFailure,
+} from '../types';
+
+export default class SMSParams {
+  public static phoneNumberToString(phoneNumber: PhoneNumber): string {
+    return `${phoneNumber.countryCode}${phoneNumber.phoneNumber}`;
+  }
+
+  public static validate(params: SendSMSParams): void {
+    /* eslint-disable-next-line max-len */
```
Comment (jjalan) [inline src/modules/communication/internals/twilio-params.ts:15]: Please fix this here and below

File: src/modules/communication/internals/twilio-params.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    const isRecipientPhoneValid = phoneUtil.isValidNumber(
+      /* eslint-disable-next-line @typescript-eslint/no-unsafe-call */
+      phoneUtil.parse(this.phoneNumberToString(params.recipientPhone)),
+    );
+
+    const isMessageValid = !!params.messageBody;
+    if (!isRecipientPhoneValid) {
+      failures.push({
+        field: 'recipientPhone',
+        message: 'Please specify valid recipient phone number.',
```
Comment (jjalan) [inline src/modules/communication/internals/twilio-params.ts:31]: specify example

File: src/modules/communication/internals/twilio-service.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  public static async sendSMS(params: SendSMSParams):Promise<void> {
+    SMSParams.validate(params);
+    try {
+      await this.twilio.messages.create({
+        to: SMSParams.phoneNumberToString(params.recipientPhone),
+        messagingServiceSid: ConfigService.getStringValue('twilio.messaging.messagingServiceSid'),
+        body: params.messageBody,
+      });
+    } catch (e) {
+      throw new ThirdPartyServiceError('SMS service unavailable.');
```
Comment (jjalan) [inline src/modules/communication/internals/twilio-service.ts:25]: Please log the third party error message so we can get notified and proactively investigate about it.

File: src/modules/communication/types.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  templateId: string;
+  templateData: LooseObject;
+};
+
+export type SendSMSParams = {
+  recipientPhone: PhoneNumber;
+  messageBody: string;
+};
+
+export enum ErrorCodes {
```
Comment (jjalan) [inline src/modules/communication/types.ts:32]: Rename it to `CommunicationErrorCode`

**Review by jjalan — CHANGES_REQUESTED** [review]

File: src/modules/communication/internals/twilio-service.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  public static async sendSMS(params: SendSMSParams):Promise<void> {
+    SMSParams.validate(params);
+    try {
+      await this.twilio.messages.create({
+        to: SMSParams.phoneNumberToString(params.recipientPhone),
+        messagingServiceSid: ConfigService.getStringValue('twilio.messaging.messagingServiceSid'),
+        body: params.messageBody,
+      });
+    } catch (e) {
+      Logger.criticial(e.message);
```
[the remark this answers is not in the document]
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/twilio-service.ts:26]: @jjalan  sir, I think ```Logger.error()``` is not available, only info, debug and warn are available.

File: src/modules/communication/internals/sendgrid-service.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      dynamicTemplateData: {
+        ...params.templateData,
+        domain: ConfigService.getStringValue('webAppHost'),
+      },
+    };
+
+    try {
+      await mail.send(msg);
+    } catch (e) {
+      Logger.criticial(e.message);
```
[the remark this answers is not in the document]
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/sendgrid-service.ts:32]: @jjalan  sir, I think ```Logger.error()``` is not available, only info, debug and warn are available.

File: src/modules/communication/internals/twilio-params.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+  public static validate(params: SendSMSParams): void {
+    const failures: ValidationFailure[] = [];
+
+    const phoneUtil = <{
+      isValidNumber: (a: string) => boolean;
+      parse: (a: string) => string
+    }>(<{
+      getInstance:() => any
+    }>PhoneNumberUtil).getInstance();
```
Comment (akashlakhera) [inline src/modules/communication/internals/twilio-params.ts:22]: Can we make it more readable?

File: src/modules/communication/internals/sendgrid-email-params.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+    if (failures.length) {
+      throw new ValidationError(
+        'Email cannot be send, please check the params validity.',
+        failures,
+      );
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
Comment (jjalan) [inline src/modules/communication/internals/sendgrid-email-params.ts:43]: Remove?

File: test/spec/email/email.spec.ts [hunk 13]
```diff
@@ -0,0 +1,180 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+import { expect } from 'chai';
+import sinon from 'sinon';
+import CommunicationServiceManager from '../../../src/modules/communication/communication-service-manager';
+import EmailService from '../../../src/modules/communication/email-service';
+import chai from 'chai';
+import chaiAsPromised from 'chai-as-promised';
```
Comment (akashlakhera) [inline test/spec/email/email.spec.ts:7]: We can also use async/await and `assert.throws` from chai. Can keep it if it's easy to use.

**Review by akashlakhera — APPROVED** [review]

File: test/spec/email/email.spec.ts (hunk 13 above)
Comment (Sparsh-Kumar) [reply test/spec/email/email.spec.ts:7]: @akashlakhera  sir, I found this an easy and readable way for test cases therefore I used it.

File: src/modules/communication/internals/sendgrid-email-params.ts (hunk 12 above)
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/sendgrid-email-params.ts:43]: yes @Sparsh-Kumar  sir, agree will make required changes.

File: src/modules/communication/internals/twilio-params.ts (hunk 11 above)
Comment (Sparsh-Kumar) [reply src/modules/communication/internals/twilio-params.ts:22]: agreed @akashlakhera  sir, will make required changes.

**Review by jjalan — APPROVED** [review]

**Pull request merged**